### PR TITLE
IGNITE-14818 Implement compressed encoding for short type in messaging.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-14818

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java [hunk 1]
```diff
@@ -236,27 +237,34 @@ public void writeBoxedByte(@Nullable Byte val) {
     /** {@inheritDoc} */
     @Override
     public void writeShort(short val) {
-        lastFinished = buf.remaining() >= 2;
+        lastFinished = buf.remaining() >= 3;
 
         if (lastFinished) {
+            val++;
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:243]: Should not this code be extracted somewhere, like `VarIntUtils`?

Comment (ibessonov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:243]: It's already copy-pasted multiple times within this class. I think I can extract the common part into a private method, would that be good?

File: modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java [hunk 2]
```diff
@@ -236,27 +237,34 @@ public void writeBoxedByte(@Nullable Byte val) {
     /** {@inheritDoc} */
     @Override
     public void writeShort(short val) {
-        lastFinished = buf.remaining() >= 2;
+        lastFinished = buf.remaining() >= 3;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:240]: Are we going to change it? In that case we need to extract the rest of magic values into constants as well. But OK, I can do that

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/InboundDecoder.java [hunk 3]
```diff
@@ -85,28 +91,43 @@ public void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
...
+
+                    if (!reader.isLastRead()) {
+                        groupTypeAttr.set(groupType);
+
+                        fixNettyBufferReaderIndex(in, buffer, initialNioBufferPosition);
+
                         break;
                     }
+
+                    msg = serializationService.createMessageDeserializer(groupType, messageType);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/netty/InboundDecoder.java:116]: Good point, I guess that's the artifact from some old implementation. I'll rename it

File: modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java [hunk 4]
```diff
@@ -237,34 +246,15 @@ public void writeBoxedByte(@Nullable Byte val) {
...
-                GridUnsafe.putByte(heapArr, baseOff + pos++, b);
-
-                intVal >>>= 7;
-            }
-
-            GridUnsafe.putByte(heapArr, baseOff + pos++, (byte) intVal);
-
-            buf.position(pos);
-        }
+        writeVarInt(Short.toUnsignedLong(++val));
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:251]:
```suggestion
        writeVarInt(Short.toUnsignedLong(val + 1));
```

File: modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java [hunk 5]
```diff
@@ -279,31 +269,15 @@ public void writeBoxedShort(@Nullable Short val) {
...
-                GridUnsafe.putByte(heapArr, baseOff + pos++, b);
-
-                val >>>= 7;
-            }
-
-            GridUnsafe.putByte(heapArr, baseOff + pos++, (byte) val);
-
-            buf.position(pos);
-        }
+        writeVarInt(Integer.toUnsignedLong(++val));
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:274]:
```suggestion
        writeVarInt(Integer.toUnsignedLong(val + 1));
```

File: modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java [hunk 6]
```diff
@@ -318,11 +292,28 @@ public void writeBoxedInt(@Nullable Integer val) {
     /** {@inheritDoc} */
     @Override
     public void writeLong(long val) {
-        lastFinished = buf.remaining() >= 10;
+        lastFinished = buf.remaining() >= MAX_VAR_LONG_BYTES;
 
-        if (lastFinished) {
-            val++;
+        writeVarInt(++val);
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java:297]:
```suggestion
        writeVarInt(val + 1);
```

**Review by rpuch — APPROVED** [review]

**Pull request merged**